src/Microsoft.AspNet.Mvc.ModelBinding/Metadata/DefaultModelMetadataProvider.cs

```diff
@@ -0,0 +1,198 @@
+﻿// Copyright (c) Microsoft Open Technologies, Inc. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Reflection;
+using Microsoft.Framework.Internal;
+
+namespace Microsoft.AspNet.Mvc.ModelBinding.Metadata
+{
+    /// <summary>
+    /// A default implementation of <see cref="IModelMetadataProvider"/> based on reflection.
+    /// </summary>
+    public class DefaultModelMetadataProvider : IModelMetadataProvider
+    {
+        private readonly TypeCache _typeCache = new TypeCache();
+        private readonly ParameterCache _parameterCache = new ParameterCache();
+        private readonly PropertiesCache _propertiesCache = new PropertiesCache();
+
+        /// <summary>
+        /// Creates a new <see cref="DefaultModelMetadataProvider"/>.
+        /// </summary>
+        /// <param name="detailsProvider">The <see cref="ICompositeMetadataDetailsProvider"/>.</param>
+        public DefaultModelMetadataProvider(ICompositeMetadataDetailsProvider detailsProvider)
+        {
+            DetailsProvider = detailsProvider;
+        }
+
+        /// <summary>
+        /// Gets the <see cref="ICompositeMetadataDetailsProvider"/>.
+        /// </summary>
+        protected ICompositeMetadataDetailsProvider DetailsProvider { get; }
+
+        /// <inheritdoc />
+        public virtual ModelMetadata GetMetadataForParameter(
+            [NotNull] ParameterInfo parameterInfo,
+            [NotNull] IEnumerable<object> attributes)
+        {
+            var key = ModelMetadataIdentity.ForParameter(parameterInfo);
+
+            DefaultMetadataDetailsCache entry;
+            if (!_parameterCache.TryGetValue(key, out entry))
+            {
+                entry = CreateParameterCacheEntry(key, attributes);
+                entry = _parameterCache.GetOrAdd(key, entry);
```

Comment:
No. This is a strictly more verbose and less performant way of using CD when the "generator" function would otherwise use a capturing lambda. I see no reason to make the other code worse.